core/src/dynamic_widget.rs

```diff
@@ -182,36 +204,113 @@ impl<D> DynRender<D> {
-        new_widgets.iter().rev().for_each(|n| {
-          cursor.insert_before(*n, arena);
-          cursor = *n;
+        let head = refresh_sign_id(*sign, &mut new_widgets, arena, store);
+        new_widgets.iter().fold(head, |prev, wid| {
+          prev.insert_after(*wid, arena);
+          wid.on_mounted_subtree(arena, store, wnd_ctx, dirty_set);
+          *wid
         });
 
-        new_widgets
-          .iter()
-          .for_each(|n| n.on_mounted_subtree(arena, store, wnd_ctx, dirty_set));
-        *siblings = new_widgets.len();
-
-        old_sign.remove_subtree(arena);
+        head.remove_subtree(arena, store);
+        *siblings = new_widgets.len()
       }
     };
+
     // Place the dynRender back in node.
     std::mem::swap(
       &mut *self.self_render.borrow_mut(),
       sign.assert_get_mut(arena),
     );
+    true
+  }
+
+  fn remove_old_subtree(
+    &self,
+    sign: WidgetId,
+    mut wid: WidgetId,
+    arena: &mut TreeArena,
+    store: &mut LayoutStore,
+    dirty_set: &DirtySet,
+  ) {
+    fn query_pin_flag(wid: WidgetId, arena: &TreeArena) -> Option<Rc<RefCell<bool>>> {
+      let mut flag = None;
+      wid
+        .assert_get(arena)
+        .query_on_first_type(QueryOrder::OutsideFirst, |w: &PinWidget| {
+          flag = Some(w.pin.clone());
+        });
+      flag
+    }
+
+    fn detach(
+      sign: WidgetId,
+      wid: WidgetId,
+      pin_flag: Rc<RefCell<bool>>,
+      arena: &mut TreeArena,
+      dirty_set: &DirtySet,
+    ) {
+      wid.detach(arena);
+      wid.assert_get(arena).query_all_type(
+        |notifier: &StateChangeNotifier| {
+          let state_changed = dirty_set.clone();
+          let pin_flag = pin_flag.clone();
+
+          // abandon the old subscribe
+          notifier.reset();
+          notifier
+            .raw_modifies()
+            .filter(|b| b.contains(ModifyScope::FRAMEWORK))
+            .subscribe(move |_| {
+              state_changed.borrow_mut().insert(wid);
+              // when unpined notify dyn widget relayout to remove self.
+              if !*pin_flag.borrow() {
+                state_changed.borrow_mut().insert(sign);
+              }
+            });
+          true
+        },
+        QueryOrder::OutsideFirst,
+      );
+    }
+
+    if wid == sign {
+      let new_sign = WidgetId::new_node(arena);
+      wid.insert_before(new_sign, arena);
+      wid.swap_id(new_sign, arena, store);
+      wid = new_sign;
+    }
+
+    let flag = query_pin_flag(wid, arena);
+    if flag.as_ref().map_or(false, |v| *v.borrow()) {
+      let flag = flag.unwrap();
+      detach(sign, wid, flag.clone(), arena, dirty_set);
+      self.pin_widgets.borrow_mut().insert(wid, flag);
+      dirty_set.borrow_mut().insert(wid);
+    } else {
+      wid.remove_subtree(arena, store);
+    }
+  }
+
+  fn destroy_unhosted(&self, arena: &mut TreeArena, store: &mut LayoutStore) {
+    let mut widgets = self.pin_widgets.borrow_mut();
+    let to_remove = widgets
```

Comment:
Needn't remove after collect, you can directly use `drain()` method.

Reply:
drain_filter is unstable